### PR TITLE
CMD-25-Filter-properties-that-form-options

### PR DESCRIPTION
**Pull request by skydeszka**

### CODE REVIEW

File: CmdStarter.Tests.Commands/Options/Attributes/IOptByAttribute.cs
Comment (djon2003) [inline CmdStarter.Tests.Commands/Options/Attributes/IOptByAttribute.cs:1]: Isn't this interface supposed to be in the `Tests.Common` project?

File: CmdStarter.Tests/Options.cs [hunk 1]
```diff
@@ -126,5 +128,42 @@ public void MethodSetup()
...
+        [TestCase<OptByExcludeAttribute>]
+        public void EnsureOptionAttributes<OptClass>() where OptClass : class, IStarterCommand, IOptByAttribute
+        {
+            starter.Namespaces = starter.Namespaces.Add(typeof(OptClass).Namespace!);
+            starter.InstantiateCommands();
+
+            var optionCommand = starter.FindCommand<OptClass>();
+            Assert.That(optionCommand, Is.Not.Null);
+
+            foreach (var includedOption in OptClass.IncludedOptions)
```
Comment (djon2003) [inline CmdStarter.Tests/Options.cs:143]: Couldn't we use `TestsCommon.AssertIEnumerablesHaveSameElements` method instead?

File: CmdStarter.Tests/Options.cs [hunk 2]
```diff
@@ -126,5 +128,42 @@ public void MethodSetup()
...
+            foreach (var excludedOption in OptClass.ExcludedOptions)
+            {
+                Assert.That(optionCommand.Options.Any(option =>
+                    option.Name.Equals(excludedOption.PascalToKebabCase())), Is.False);
+            }
+        }
+
+        [TestCase<OptAllPropertiesExcluded>(OptAllPropertiesExcluded.OPTION_COUNT)]
+        [TestCase<OptAllPropertiesExcludedWithInheritance>(OptAllPropertiesExcludedWithInheritance.OPTION_COUNT)]
+        public void EnsureClassExcludeAttribute<OptClass>(int optionCount) where OptClass : class, IStarterCommand
```
Comment (djon2003) [inline CmdStarter.Tests/Options.cs:158]: This is good, but a simple question. Why not using the new test method above?

File: CmdStarter.Tests.Commands/Options/Attributes/OptByAttribute.cs
Comment (djon2003) [inline CmdStarter.Tests.Commands/Options/Attributes/OptByAttribute.cs:1]: I know it is really trivial and in fact probably already tested all over, but not formerly. The initial case: no attribute that transposes all properties into options. Shouldn't we have one?

**Review by djon2003 — CHANGES_REQUESTED** [review]

File: CmdStarter.Tests.Commands/Options/Attributes/IOptByAttribute.cs
Comment (skydeszka) [reply CmdStarter.Tests.Commands/Options/Attributes/IOptByAttribute.cs:1]: Done

File: CmdStarter.Tests/Options.cs (hunk 1 above)
Comment (skydeszka) [reply CmdStarter.Tests/Options.cs:144]: Done

File: CmdStarter.Tests/Options.cs (hunk 2 above)
Comment (skydeszka) [reply CmdStarter.Tests/Options.cs:159]: What do you mean exactly?

Why I didn't do this in that method or why I didn't copy a logic from there?

Comment (djon2003) [reply CmdStarter.Tests/Options.cs:159]: I mean: why those test cases aren't tested using `EnsureOptionAttributes` ?

File: CmdStarter.Tests.Commands/Options/Attributes/OptByAttribute.cs
Comment (skydeszka) [reply CmdStarter.Tests.Commands/Options/Attributes/OptByAttribute.cs:1]: Done

File: CmdStarter.Tests/Options.cs (hunk 2 above)
Comment (skydeszka) [reply CmdStarter.Tests/Options.cs:159]: Oh, I see!

Well... good question hehe

Comment (skydeszka) [reply CmdStarter.Tests/Options.cs]: Done

File: CmdStarter.Tests/Options.cs [hunk 3]
```diff
@@ -141,11 +141,9 @@ public void MethodSetup()
...
-            foreach (var includedOption in OptClass.IncludedOptions)
-            {
-                Assert.That(optionCommand.Options.All(option =>
-                    option.Name.Equals(includedOption.PascalToKebabCase())), Is.True);
-            }
+            TestsCommon.AssertIEnumerablesHaveSameElements(
+                optionCommand.Options.Select(o => o.Name),
+                OptClass.IncludedOptions.Select(o => o.PascalToKebabCase()));
 
             foreach (var excludedOption in OptClass.ExcludedOptions)
```
[the remark this answers is not in the document]
Comment (skydeszka) [reply CmdStarter.Tests/Options.cs:151]: Not really, it checks if the IEnumerable has any elements of the other, not that they have the same elements of not.

[the remark this answers is not in the document]
Comment (djon2003) [reply CmdStarter.Tests/Options.cs:151]: Oups! You are right. We'll keep it like this.

**Review by djon2003 — APPROVED** [review]

**Pull request merged**